### PR TITLE
Add the Rails Type Casting talk proposal

### PR DESCRIPTION
**Pull request by sgrif**

### CODE REVIEW

**Comment by iwz** [comment]
I'd at least start with "Type casting in Active Record" as the title

File: rails-type-casting.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+Title
+==
+
+I don't know I'm bad at naming things
```
Comment (jferris) [inline rails-type-casting.md]: This is a great title for a type casting talk.

Comment (iwz) [inline rails-type-casting.md]: lol

File: rails-type-casting.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+Finally, we'll look at the new `attributes` API, which is coming in Rails 5.
+It's one of several APIs being introduced which are meant to give greater, lower
+level, control over Active Record, and turn off some of the "magic". We'll look
+at why this is a good thing, and how building these simple APIs allows you to
+compose really powerful high level APIs with only a few lines of code.
+
+The intended audience take-away is primarily geared towards the design of APIs,
+and why simple is often better. You can build powerful, complex APIs by
+composing simple pieces, and there is a ton of value in making those smaller
+pieces accessible as well, like we're doing in Rails 5.
```
Comment (mike-burns) [inline rails-type-casting.md]: This paragraph sounds awesome and I'd go to a talk that mentioned this in the abstract.

Comment (r00k) [inline rails-type-casting.md]: I agree with Mike. I think this is the most compelling part of the proposal.

**Comment by r00k** [comment]
Personal opinion: 

"how ActiveRecord typecasts things" #=> ZZZZzzzz

"how to design a great API, with a real-world example from an important open-source project" #=> I'd go to that.

Also, I'd cut 50% of the words and spend some time choosing a provocative title.

**Comment by sgrif** [comment]
Yeah, this is definitely intended to be primarily about process, with the type casting system being the side effect. I'll see if I can revise to make that clearer.

File: rails-type-casting.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+==
+
+Type casting in Active Record is an integral part of Rails, but very few of us
+know how it works and the changes it has gone through. We're going to unravel
+the mysteries behind how Active Record type casts anything you give to it. We'll
+look at the major changes that happened in Rails 4.2, and show the new API
+coming in Rails 5. We'll be discussing the process of cleaning up complex legacy
+code. Finally, we'll look at the power of simple, composeable APIs, and how this
+affects the future of Rails.
+
```
Comment (derekprior) [inline rails-type-casting.md]: The last sentence hints at why I, as a mere Rails Developer, want to go to this talk. The refactoring is interesting to me, but it's not immediately clear what I'll get out of it.

File: rails-type-casting.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+The talk will start with a brief walk through of what type casting is, and when
+Active Record will type cast your input. We'll look at what the code looked like
+in Rails 4.1, and what it looks like in Rails 4.2. The main focus of the talk is
+process.
+
+Many people look at the Rails code base and assume that it's too clever, or
+they're too inexperienced to understand it. In reality, Rails just a really
+large legacy code base. The same challenges and precesses apply here as it would
+to anyone's legacy program.
```
Comment (derekprior) [inline rails-type-casting.md]: Okay, we're on to something here. "Rails as a large, legacy code base..."

File: rails-type-casting.md (hunk 2 above)
Comment (derekprior) [inline rails-type-casting.md]: Third! Hit part of this (and Rails as a legacy code base, I think) in the abstract. That's what people are going to see in order to decide to go to your talk.

File: rails-type-casting.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+compose really powerful high level APIs with only a few lines of code.
+
+The intended audience take-away is primarily geared towards the design of APIs,
+and why simple is often better. You can build powerful, complex APIs by
+composing simple pieces, and there is a ton of value in making those smaller
+pieces accessible as well, like we're doing in Rails 5.
+
+After this talk the audience will have a better understanding of a part of
+Active Record that affects every application, and can use that knowledge to
+improve their apps.
```
Comment (derekprior) [inline rails-type-casting.md]: Maybe. I think the take away above is a lot more interesting.

**Comment by sgrif** [comment]
Ready for re-review if anyone has time.

File: rails-type-casting.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+I don't know I'm bad at naming things
+
+Abstract
+==
+
+The most useful APIs are simple and composeable. Omnipotent DSLs can be great,
+but sometimes we want to just write Ruby. We're going to look at the process of
+designing a new API for attributes and type casting in Rails 5.0, and why
+simpler is better. We'll unravel some of the mysteries behind the internals
+After all, Rails is ultimately just a large legacy code base.
```
Comment (derekprior) [inline rails-type-casting.md]: OMG YES. Now I'm going.

File: rails-type-casting.md (hunk 1 above)
Comment (sgrif) [inline rails-type-casting.md]: No but seriously does anyone have any suggestions for this?

File: rails-type-casting.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+I don't know I'm bad at naming things
+
+Abstract
+==
+
+The most useful APIs are simple and composeable. Omnipotent DSLs can be great,
+but sometimes we want to just write Ruby. We're going to look at the process of
+designing a new API for attributes and type casting in Rails 5.0, and why
+simpler is better. We'll unravel some of the mysteries behind the internals
```
Comment (tessgriffin) [inline rails-type-casting.md]: You need a period at the end of this line.

File: rails-type-casting.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+used to tease them out slowly. And you'll be getting a sneak peak at several of
+the features coming in Rails 5.0.
+
+Details
+==
+
+The talk will start with a brief walk through of what type casting is, and when
+Active Record will type cast your input. Then we'll look at the API we wanted to
+have to hook into this process. The focus for the design of the attributes API
+was simplicity, and composeability. It quickly became apparent just how many
```
Comment (tessgriffin) [inline rails-type-casting.md]: You shouldn't need a comma here between simplicity and compseability.

File: rails-type-casting.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+Many people look at the Rails code base and assume that it's too clever, or
+they're too inexperienced to understand it. In reality, Rails just a really
+large legacy code base. The same challenges and precesses apply here as it would
+to anyone's legacy program.
+
+The talk will look at how important test coverage is, and how we can break down
+a large sweeping change into trivial, small steps. We'll look at how good object
+oriented design can lead to code which is much easier to test, and much easier
+to get feedback from your test failures.
```
Comment (tessgriffin) [inline rails-type-casting.md]: I don't think you need a comma here.

File: rails-type-casting.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+This talk involves major changes to a core part of Active Record, intended to
+make it accessible by every user of Rails. We will also be looking at new APIs
+that will be available for public consumption in Rails 5, which will make
+several common patterns much easier, and give much more control over how to
+convert custom objects to and from SQL.
+
+What is your involvement?
+==
+
+I was the Rails contributor responsible for the majority of these changes
```
Comment (tessgriffin) [inline rails-type-casting.md]: Should probably add a period here.

File: rails-type-casting.md (hunk 1 above)
Comment (tessgriffin) [inline rails-type-casting.md]: Maybe something along the lines of "You don't need magic to write Ruby" or something like that.

**Pull request merged**